snuba/consumers/consumer_builder.py

```diff
@@ -270,12 +273,22 @@ def __build_streaming_strategy_factory(
 
         return strategy_factory
 
-    def close_producers(self) -> None:
-        if self.commit_log_producer is not None:
-            self.commit_log_producer.close()
+    def close(self, consumer: StreamProcessor[KafkaPayload]) -> None:
+        if not self.commit_log_producer and not self.replacements_producer:
+            consumer.run()
```

Comment:
Wait... why did this move in here? Now your close function is not just closing, it's actually responsible for running the consumer too. That doesn't seem right.

Reply:
This will change now since we realized that this type of Producer can't be closed. I had it like this originally because the usages of `closing()` I found in the codebase all did it like something similar to this:

```if dead_letter_producer:
        with closing(dead_letter_producer):
            processor.run()
    else:
        processor.run()```

Reply:
I think @ayirr7  is doing this because, after creating the consumer she does not have access to the producers anymore from the code that calls the consumer builder and the consumer builder is gone. 

This highlights a larger problem we have with resources life cycle in this case and in Arroyo in general. 
- Arroyo is not strongly opinionated. It allows you to pass resources in (like producers) to factories or consumers and retain the ownership but it also allows you to give the ownership entirely to the consumer. The processing factory creates and destroys what it needs. The second approach is troublesome because it makes unit tests much harder so should be avoided. 
- The consumer builder is not built with resource management in mind. Cannot remember if this was intentional or a mistake, but definitively it does not work with Arroyo resources management. 

It seems that the consumer builder would work in a world where it creates the resources (the producer) pass the ownership to the consumer and let the consumer figure out when to close them. This is what we generally avoid in Arroyo.

I think the options you are are these (they are all bad):
- Make the consumer builder return a new class that retains ownership of the consumer and of the producers. This class has a run method and a close method. When you call close it closes the producers. So you preserve the lifecycle of the builder, which is not supposed to survive after it has built the consumer (which is cumbersome and full of boiler plate)
- You make the consumer builder return both producers and consumer (which is gross). Please let's avoid.
- We get rid of the consumer builder entirely. But I feel that is slightly beyond the scope of this PR.
- You instantiate the producer in the consumer builder, pass them to the consumer and the batch writer and rely on them to close the producer. This is confusing as it splits the life cycle so hard to verify, though possible with the `KafkaConsumerWithCommitLog`. Not sure about the batch writer.

I would avoid making the consumer instantiate its own producers as it makes it untestable.

Reply:
@fpacifici I think you're referring to a somewhat related but different problem. To clarify I think this was fine before the latest change in https://github.com/getsentry/snuba/pull/3305/commits/392db09886d3bf768fe4f52be65f01f9a0a6b6e5 which started passing a consumer in here. I think we should just revert that last part and call it a day.

Reply:
Could we at least call that method `run_and_close` ? A `close` method that also runs all the business logic would be counter intuitive.

Reply:
Running went back to it's original place now.